rustls/src/crypto/ring/tls13.rs

```diff
@@ -0,0 +1,224 @@
+        #[cfg(feature = "quic")]
+        confidentiality_limit: 1 << 23,
+        #[cfg(feature = "quic")]
+        integrity_limit: 1 << 52,
+    });
+
+/// The TLS1.3 ciphersuite TLS_AES_128_GCM_SHA256
+pub static TLS13_AES_128_GCM_SHA256: SupportedCipherSuite =
+    SupportedCipherSuite::Tls13(TLS13_AES_128_GCM_SHA256_INTERNAL);
+
+pub(crate) static TLS13_AES_128_GCM_SHA256_INTERNAL: &Tls13CipherSuite = &Tls13CipherSuite {
+    common: CipherSuiteCommon {
+        suite: CipherSuite::TLS13_AES_128_GCM_SHA256,
+        bulk: BulkAlgorithm::Aes128Gcm,
+        hash_provider: &super::hash::SHA256,
+    },
+    hmac_provider: &super::hmac::HMAC_SHA256,
+    aead_alg: &AeadAes128Gcm(RingAeadFactory(&ring::aead::AES_128_GCM)),
+    #[cfg(feature = "quic")]
+    confidentiality_limit: 1 << 23,
+    #[cfg(feature = "quic")]
+    integrity_limit: 1 << 52,
+};
+
+struct Tls13MessageEncrypter {
+    enc_key: ring::aead::LessSafeKey,
+    iv: Iv,
+}
+
+struct Tls13MessageDecrypter {
+    dec_key: ring::aead::LessSafeKey,
+    iv: Iv,
+}
+
+impl MessageEncrypter for Tls13MessageEncrypter {
+    fn encrypt(&self, msg: BorrowedPlainMessage, seq: u64) -> Result<OpaqueMessage, Error> {
+        let total_len = msg.payload.len() + 1 + self.enc_key.algorithm().tag_len();
+        let mut payload = Vec::with_capacity(total_len);
+        payload.extend_from_slice(msg.payload);
+        msg.typ.encode(&mut payload);
+
+        let nonce = aead::Nonce::assume_unique_for_key(make_nonce(&self.iv, seq));
+        let aad = aead::Aad::from(make_tls13_aad(total_len));
+
+        self.enc_key
+            .seal_in_place_append_tag(nonce, aad, &mut payload)
+            .map_err(|_| Error::General("encrypt failed".to_string()))?;
+
+        Ok(OpaqueMessage {
+            typ: ContentType::ApplicationData,
+            version: ProtocolVersion::TLSv1_2,
+            payload: Payload::new(payload),
+        })
+    }
+}
+
+impl MessageDecrypter for Tls13MessageDecrypter {
+    fn decrypt(&self, mut msg: OpaqueMessage, seq: u64) -> Result<PlainMessage, Error> {
+        let payload = &mut msg.payload.0;
+        if payload.len() < self.dec_key.algorithm().tag_len() {
+            return Err(Error::DecryptError);
+        }
+
+        let nonce = aead::Nonce::assume_unique_for_key(make_nonce(&self.iv, seq));
+        let aad = aead::Aad::from(make_tls13_aad(payload.len()));
+        let plain_len = self
+            .dec_key
+            .open_in_place(nonce, aad, payload)
+            .map_err(|_| Error::DecryptError)?
+            .len();
+
+        payload.truncate(plain_len);
+        self.tls13_check_length_and_unpad(msg)
+    }
+}
+
+struct RingAeadFactory(&'static ring::aead::Algorithm);
+
+impl RingAeadFactory {
+    fn encrypter(&self, key: AeadKey, iv: Iv) -> Box<dyn MessageEncrypter> {
+        Box::new(Tls13MessageEncrypter {
+            enc_key: aead::LessSafeKey::new(aead::UnboundKey::new(self.0, key.as_ref()).unwrap()),
```

Comment:
same comment as in an earlier commit: perhaps worth a comment about the safety of `unwrap` w.r.t an externally enforced invariant?